iot-hub/Samples/device/DeviceReconnectionSample/DeviceReconnectionSample.cs

```diff
@@ -122,11 +115,18 @@ private void InitializeClient()
             }
         }
 
-        private void ConnectionStatusChangeHandler(ConnectionStatus status, ConnectionStatusChangeReason reason)
+        // It is not good practice to have async void methods, however, DeviceClient.SetConnectionStatusChangesHandler() event handler signature has a void return type.
+        // As a result, any operation within this block will be executed unmonitored on another thread.
+        // To prevent multi-threaded synchronization issues, the async method InitializeClientAsync being called in here first grabs a lock
+        // before attempting to initailize or dispose the device client instance.
+        private async void ConnectionStatusChangeHandler(ConnectionStatus status, ConnectionStatusChangeReason reason)
         {
-            _logger.LogDebug($"Connection status changed: status={status}, reason={reason}");
+            lock (_connectionStatusLock)
+            {
+                _logger.LogDebug($"Connection status changed: status={status}, reason={reason}");
+                s_connectionStatus = status;
```

Comment:
If you are locking to avoid it being overwritten in another thread, you still have the problem of the value changing from here to the switch statement.

My feeling is there is no value in locking here. Just use volatile, or in the switch statement use `status` not the member variable.

Reply:
Yeah, I was counting on it being fast enough to not matter, but that will not be true for all cases. I could put the entire switch block inside the lock, but that seemed to be an overkill. I like the idea of using the method arg `status` directly in the switch, I'll do that.

Reply:
If you are expecting multiple threads to access this, it could cause lots of other issues. Ex: You are trying to initialize the client inside the switch but there is another thread that sets the s_connectionStatus  to something else, causing the init to skip completely.

Reply:
but that is what we want, we don't initialize s_connectionStatus anywhere other than within this callback, so if the callback updated s_connectionStatus to be Connected, we want the init block to be skipped. For this reason, we have a check for the status s_connectionStatus inside the init block as well.

Reply:
And if threads can re-write this value in any order, we might be losing some info on s_connectionStatus. It should be an OR or AND based on the logic we use to store the result of all callbacks.

Reply:
Talked offline